### PR TITLE
rewrite MatrixExplodeRows using compiled IR functions; add tests

### PR DESCRIPTION
**Pull request by catoverdrive**

### CODE REVIEW

**Comment by danking** [comment]
I'll do this tomorrow, sorry for the delay

File: src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 1]
```diff
@@ -1899,47 +1899,64 @@ case class MatrixExplodeRows(child: MatrixIR, path: IndexedSeq[String]) extends
...
+        val lenF = l()
+        val rowF = f()
         it.flatMap { rv =>
-          ur.set(rv)
-          val keys = localQuerier(ur).asInstanceOf[Iterable[Any]]
-          if (keys == null)
-            None
+          val len = lenF(rv.region, rv.offset, false)
+          if (len == 0)
+            Iterator()
```
Comment (danking) [inline src/main/scala/is/hail/expr/ir/MatrixIR.scala]: `Iterator.empty` avoids an allocation (or two?)

Comment (danking) [reply src/main/scala/is/hail/expr/ir/MatrixIR.scala]: not sure if this extra if and all is really necessary anyway, the iterator below handles `len == 0` fine.

Comment (danking) [reply src/main/scala/is/hail/expr/ir/MatrixIR.scala]: You've deftly collapsed the `len == 0` and the `null` case in your definition of `lenF`

File: src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala [hunk 2]
```diff
@@ -117,5 +118,30 @@ class MatrixIRSuite extends SparkSuite {
...
+  @Test def testMatrixExplode() {
+    val tarray = TArray(TInt32())
+    val range = MatrixTable.range(hc, 5, 2, None).ast
+    val fields = FastIndexedSeq(
+      "empty" -> IRArray(),
+      "null" -> NA(tarray),
+      "set" -> IRSet(1, 3),
+      "z" -> IRArray(3),
+      "x" -> IRStruct("y" -> IRArray(3)),
+      "a" -> IRStruct("b" -> IRStruct("c" -> IRArray())))
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: can we add one field which is >1 nested and has more than one element?

File: src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala [hunk 3]
```diff
@@ -117,5 +118,30 @@ class MatrixIRSuite extends SparkSuite {
     val actualOrdering = getRows(unioned).map { case Row(i: Int) => i }
 
     assert(actualOrdering sameElements expectedOrdering)
+
+  }
+
+  @Test def testMatrixExplode() {
```
Comment (danking) [inline src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: I feel like we should generally be moving towards python tests. It'd also be easy there to assert that collecting the exploded field has the expected contents in the expected order.

Comment (catoverdrive) [reply src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: python tests already exist. I'm adding scala tests for this issue:
https://github.com/hail-is/hail/issues/4016

File: src/main/scala/is/hail/expr/ir/MatrixIR.scala (hunk 1 above)
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/MatrixIR.scala]: yeah, I thought I took this out :( I'll take it out.

File: src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 4]
```diff
@@ -1899,47 +1899,64 @@ case class MatrixExplodeRows(child: MatrixIR, path: IndexedSeq[String]) extends
...
   }
 
-  private val (newRVType, inserter) = rvRowType.unsafeStructInsert(keyType, path.toList)
+  val idx = Ref(genUID(), TInt32())
+  val newRVRow: InsertFields = {
+    val refs = path.init.scanLeft(Ref("va", rvRowType))((struct, name) =>
+      Ref(genUID(), coerce[TStruct](struct.typ).field(name).typ))
 
-  val typ: MatrixType = child.typ.copy(rvRowType = newRVType)
+    path.zip(refs).zipWithIndex.foldRight[IR](idx) {
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/ir/MatrixIR.scala]: I'll put one in.

File: src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala (hunk 3 above)
Comment (danking) [reply src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: What's the motivation for #4016 cc: @cseed?

**Pull request merged**